File: backend/enrollment-service/src/main/java/com/grammarcetamol/enrollment/client/CourseServiceClient.java
package com.grammarcetamol.enrollment.client;

import com.grammarcetamol.enrollment.config.AppProperties;
import com.grammarcetamol.shared.dto.ApiResponse;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.client.SimpleClientHttpRequestFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestClient;

import java.util.UUID;

/**
 * Plain internal REST call to course-service — no gRPC (course-service has no gRPC
 * infrastructure; adding it just for this would be more new surface area than a direct call to
 * an already-public endpoint). Never goes through the gateway.
 *
 * course-service's authorization model only reveals non-preview lesson videoUrl / non-published
 * courses to the owning admin/moderator (see CurrentUser.canModify). Enrollment Service needs the
 * full curriculum (including video URLs) for any enrolled student, not just the course owner —
 * there's no "enrolled student" concept on course-service's side to check against. So this client
 * presents as a trusted internal caller (X-User-Role: SUPER_ADMIN) rather than forwarding the real
 * end user's identity. This is safe because course-service is never reachable directly from a
 * browser (only from other backend services on the internal network) — the same implicit trust
 * boundary every header-trust service already relies on for X-User-Id/X-User-Role.
 */
@Component
public class CourseServiceClient {

    private static final UUID INTERNAL_CALLER_ID = new UUID(0, 0);

    private final RestClient restClient;

    public CourseServiceClient(AppProperties appProperties) {
        // SimpleClientHttpRequestFactory (java.net.HttpURLConnection) instead of RestClient's
        // default java.net.http.HttpClient — the latter opens an NIO Selector on construction,
        // which trips the same pre-existing Windows loopback-socket issue documented in
        // auth-service/course-service's READMEs (security software intercepting the loopback
        // socket). Classic blocking I/O sidesteps that entirely, which is a real reliability win
        // on affected dev machines, not just a workaround for automated verification.
        this.restClient = RestClient.builder()
            .baseUrl(appProperties.getCourseServiceUrl())
            .requestFactory(new SimpleClientHttpRequestFactory())
            .defaultHeader("X-User-Id", INTERNAL_CALLER_ID.toString())
            .defaultHeader("X-User-Role", "SUPER_ADMIN")
            .build();
    }

    public CourseDetailDto getCourse(UUID courseId) {
        try {
            ApiResponse<CourseDetailDto> response = restClient.get()
                .uri("/api/courses/{id}", courseId)
                .retrieve()
                .body(new ParameterizedTypeReference<>() {});
            if (response == null || response.getData() == null) {
                throw new EntityNotFoundException("Course not found: " + courseId);
            }
            return response.getData();
        } catch (HttpClientErrorException.NotFound e) {
            throw new EntityNotFoundException("Course not found: " + courseId);
        }
    }
}

File: backend/upload-service/src/course-client/course-service.client.ts
import {Injectable, NotFoundException, ServiceUnavailableException} from '@nestjs/common';
import {ConfigService} from '@nestjs/config';

export interface CourseSummary {
  id: string;
  title: string;
  slug: string;
  status: string;
  price: string;
  currency: string;
}

interface CourseApiResponse {
  success: boolean;
  data?: { course: CourseSummary };
  error?: string;
}

// Same internal-caller convention as payment-service's CourseServiceClient — never goes through
// the gateway, presents as a trusted SUPER_ADMIN caller since course-service has no concept of
// "this is an in-progress upload session" to authorize against otherwise. Used only to validate
// the course exists and to get its title server-side — never trust a client-supplied course name.
const INTERNAL_CALLER_ID = '00000000-0000-0000-0000-000000000000';

@Injectable()
export class CourseServiceClient {
  constructor(private readonly config: ConfigService) {}

  async getCourse(courseId: string): Promise<CourseSummary> {
    const baseUrl = this.config.get<string>('COURSE_SERVICE_URL', 'http://localhost:9002');
    const res = await fetch(`${baseUrl}/api/courses/${courseId}`, {
      headers: {
        'X-User-Id': INTERNAL_CALLER_ID,
        'X-User-Role': 'SUPER_ADMIN',
      },
    });

    if (res.status === 404) {
      throw new NotFoundException(`Course not found: ${courseId}`);
    }
    if (!res.ok) {
      throw new ServiceUnavailableException('course-service request failed');
    }

    const body = (await res.json()) as CourseApiResponse;
    if (!body.success || !body.data) {
      throw new NotFoundException(`Course not found: ${courseId}`);
    }
    return body.data.course;
  }
}
